### PR TITLE
[4.1] Parameters for Module Chromes

### PR DESCRIPTION
**Pull request by Bakual**
This isn't ready to be used yet as it needs some discussion first.

### Summary of Changes
All it does is looking in the chrome jlayout folders for XML files and loading them into the module edit form.

### Testing Instructions
Create an XML in a chrome folder and put some fields into it. As an example create the file `templates/cassiopeia/html/layouts/chromes/params.xml` and put the following content into it:
````
<?xml version="1.0" encoding="utf-8"?>
<form>
	<fields name="params">
		<fieldset
				name="advanced">

			<field
					name="cassiopeia_test"
					type="text"
					label="Test Field Cassiopeia"
					showon="style:Cassiopeia-card"
			/>

		</fieldset>
	</fields>
</form>
````

### Expected result
The defined parameter should be available in the module edit form.
Since the example has a `showon` defined, it will only show up if you have selected the module style "card" from Cassiopeia.

image

### Documentation Changes Required
Obviously.

### CODE REVIEW

**Comment by Bakual** [comment]
@ciar4n mentioned in https://github.com/joomla/joomla-cms/pull/23570#issuecomment-455163103 the fact that the way it is done here means template devs need to play nice. The showon would be essential and I also think the prefixed field name.

There are some other thoughts which are needed here. Some that come to mind:
* What happens if the module style is left on `inherited`. Currently the custom chrome parameters aren't shown the way the showon is written. Would that be acceptable?
* What should we do with the existing default chrome fields as defined in https://github.com/joomla/joomla-cms/blob/4.0-dev/administrator/components/com_modules/forms/advanced.xml
* Current code enforces no naming convention for the XML. You can have one file per template were you define all params in one place or one file per chrome or even more. Allowing multiple files would allow the user to add more parameters for his own custom chromes.
* What should happen if we have multiple fields with the same name but different attributes.

**Comment by ciar4n** [comment]
> What happens if the module style is left on inherited. Currently the custom chrome parameters aren't shown the way the showon is written. Would that be acceptable?

Tricky. I guess this means that to show the parameters, a user has to select the chrome from the list, even thou the selected chrome is the same as the inherited. Seems incorrect but not sure how to work around it.

**Comment by Bakual** [comment]
> Seems incorrect but not sure how to work around it.

Yep, I have the same feeling. Imho we can either show all possible parameters, only the default ones or none. We have no way of knowing which is the inherited style, and it may even change when the active template changes.

**Comment by ciar4n** [comment]
Personally, I would lean towards displaying none. Assuming that can be enforced and the template dev can't override it by simply not setting the 'showon'.

**Comment by roland-d** [comment]
@Bakual Any interest in picking this up and fixing the conflicts?

**Comment by Bakual** [comment]
Yep, definitively interested, but certainly not in the comins week or two. Maybe later I can fix it,
Since it's an RFC, I would be interested in more opinions on the matter anyway.

**Comment by Bakual** [comment]
Fixed the conflicts (they were coming from the now merged other PR). Adjusted the PR description.

**Comment by punambaravkar** [comment]
I have tested this item :white_check_mark: successfully on 67d4641d2335c29f239f1c48d7b382f8cae90745<br /><br />Thank you for adding Parameters for Module Chromes<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23577">issues.joomla.org/tracker/joomla-cms/23577</a>.</sub>

**Comment by priyankaSahutekdi** [comment]
I have tested this item :white_check_mark: successfully on 67d4641d2335c29f239f1c48d7b382f8cae90745<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23577">issues.joomla.org/tracker/joomla-cms/23577</a>.</sub>

**Comment by HLeithner** [comment]
@Bakual can you merge the current 4.1-dev into it and change the target branch because this is a new feature and it's too late for j4.0.

**Comment by Bakual** [comment]
PR is rebased on 4.1-dev now. Code has not changed.

**Comment by Bakual** [comment]
Removed the use of JForm in favor of Form 😄 
Now skipping a template if there is no chrome folder in it (was giving an warning message before)

**Comment by chmst** [comment]
I like this feature, and it works as described. 
As it is a new feature, and documentation is needed, we could do a reordering of the fields, so that it is better understandable? 

At least move the field module style at the end, so it is together with newly pluggend in params? @Bakual

**Comment by Bakual** [comment]
That's a good idea. I moved the style parameter to the end so it's next to the chrome parameters.

File: administrator/components/com_modules/src/Model/ModuleModel.php [hunk 1]
```diff
@@ -900,6 +902,46 @@ protected function preprocessForm(Form $form, $data, $group = 'content')
...
+				$form->loadFile(basename($formFile, '.xml'), false);
+			}
+		}
+
+		// Load chrome specific params for template files
+		$templates = ModulesHelper::getTemplates($clientId);
+
+		foreach ($templates as $template)
+		{
+			$chromePath      = $client->path . '/templates/' . $template->element . '/html/layouts/chromes';
```
Comment (richard67) [inline administrator/components/com_modules/src/Model/ModuleModel.php]:
```suggestion
			$chromePath = $client->path . '/templates/' . $template->element . '/html/layouts/chromes';
```

**Comment by chmst** [comment]
I have tested this item :white_check_mark: successfully on 71710447df66ff1d67ca01e67271db1d24c9ce6b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23577">issues.joomla.org/tracker/joomla-cms/23577</a>.</sub>

**Comment by RickR2H** [comment]
I have tested this item :white_check_mark: successfully on 71710447df66ff1d67ca01e67271db1d24c9ce6b<br /><br />Field shows up and works and parameter is stored in the DB.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23577">issues.joomla.org/tracker/joomla-cms/23577</a>.</sub>

**Comment by RickR2H** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23577">issues.joomla.org/tracker/joomla-cms/23577</a>.</sub>

**Comment by bembelimen** [comment]
Thx

**Comment by RickR2H** [comment]
@Bakual I just noticed that language strings is the params.xml are not working. In your example you use hard coded language strings. It would be nice if language strings can be added so the parameters become Multilingual. Maybe add a new PR. Ping me if you need someone to test :wink: <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23577">issues.joomla.org/tracker/joomla-cms/23577</a>.</sub>

**Comment by Bakual** [comment]
Language strings work. I actually just used the feature this week. Did you put the language strings into the templates `template_name.sys.ini` language file? Because I think only that one is loaded.

**Comment by RickR2H** [comment]
@Bakual Looks like I added the translations to the other file without the sys. Thanks for the tip!

**Pull request merged**